tests/test_class2D.py

```diff
@@ -0,0 +1,330 @@
+        Trivial test of rotation in FSPCA Basis.
+
+        Also covers to_real and to_complex conversions in FSPCA Basis.
+        """
+        coef = self.fspca_basis.expand_from_image_basis(self.imgs)
+        # rotate by pi
+        rot_coef = self.fspca_basis.rotate(coef, radians=np.pi)
+        rot_imgs = self.fspca_basis.evaluate_to_image_basis(rot_coef)
+
+        for i, img in enumerate(self.imgs):
+            rmse = np.sqrt(np.mean(np.square(np.flip(img) - rot_imgs[i])))
+            self.assertTrue(rmse < 10 * utest_tolerance(self.dtype))
+
+
+class RIRClass2DTestCase(TestCase):
+    def setUp(self):
+        self.resolution = 16
+        self.dtype = np.float64
+
+        # Create some projections
+        v = Volume(
+            np.load(os.path.join(DATA_DIR, "clean70SRibosome_vol.npy")).astype(
+                self.dtype
+            )
+        )
+        v = v.downsample(self.resolution)
+
+        # Clean
+        self.clean_src = Simulation(
+            L=self.resolution,
+            n=321,
+            vols=v,
+            dtype=self.dtype,
+        )
+
+        # With Noise
+        noise_var = 0.01 * np.var(np.sum(v[0], axis=0))
+        noise_filter = ScalarFilter(dim=2, value=noise_var)
+        self.noisy_src = Simulation(
+            L=self.resolution,
+            n=123,
+            vols=v,
+            dtype=self.dtype,
+            noise_filter=noise_filter,
+        )
+
+        # Set up FFB
+        # Setup a Basis
+        self.basis = FFBBasis2D((self.resolution, self.resolution), dtype=self.dtype)
+        # Calculate Fourier Bessel Coefs
+        self.clean_coefs = self.basis.evaluate_t(
+            self.clean_src.images(0, self.clean_src.n)
+        )
+        self.noisy_coefs = self.basis.evaluate_t(
+            self.noisy_src.images(0, self.clean_src.n)
+        )
+
+        # Create Basis
+        self.clean_fspca_basis = FSPCABasis(
+            self.clean_src, self.basis, noise_var=0
+        )  # Note noise_var assigned zero, skips eigval filtering.
+        self.clean_fspca_basis.build(self.clean_coefs)
+
+        self.noisy_fspca_basis = FSPCABasis(self.noisy_src, self.basis)
+        self.noisy_fspca_basis.build(self.noisy_coefs)
+
+    def testClass2DBase(self):
+        """
+        Make sure the base class doesn't crash when using arguments.
+        """
+        _ = Class2D(self.clean_src)  # Default dtype
+        _ = Class2D(self.clean_src, dtype=self.dtype)  # Consistent dtype
+        _ = Class2D(self.clean_src, dtype=np.float16)  # Different dtype
+
+    def testRIRLegacy(self):
+        """
+        Currently just tests for runtime errors.
+        """
+        rir = RIRClass2D(
+            self.clean_src,
+            self.clean_fspca_basis,
+            large_pca_implementation="legacy",
+            nn_implementation="legacy",
+            bispectrum_implementation="legacy",
+        )
+
+        result = rir.classify()
+        _ = rir.output(*result[:3], include_refl=False)
```

Comment:
`_legacy_pca` is tested under `testRIRLegacy`.  If you mean in conjunction with `bispectrum_implementation="devel"`, then no its not.  

I added a smoke test, but I'm not sure it addes much value for the additional time.

My goal wasn't to test every combination, rather each module should get smoke tested at least once if reasonable to do so... thus avoiding calling slow stuff too many times.

Will need to create some larger "slow" flagged unit tests once we're reasonably happy with the results. (Mainly because that end result is what we would need to generate acceptable error intervals in practice).

Reply:
`self.pca` is only run under `bispectrum_implementation="devel"`, so `testRIRLegacy` doesn't end up running `_legacy_pca`. I noticed because I was trying to print out some debug values there and nothing happened.

I'm not saying we need to test every combination. Just that this particular one is needed for `_legacy_pca` to be covered by tests.